### PR TITLE
#1929 Implemented a search functionality for adding a new group

### PR DESCRIPTION
**Pull request by ebouJ**
I implemented a Search functionality to search for users when creating a new group. 

simulator screen shot - iphone 6 - 2018-03-24 at 04 39 33

### CODE REVIEW

File: src/group/GroupCard.js [hunk 1]
```diff
@@ -103,7 +104,7 @@ export default class GroupCard extends PureComponent<Props, State> {
           style={styles.list}
           ownEmail={ownEmail}
           filter=""
```
Comment (jainkuniya) [inline src/group/GroupCard.js]: Pass filter to it.
like https://github.com/zulip/zulip-mobile/blob/d1b36684a6f99955699b1a4551d98be1333742db/src/users/UsersCard.js#L31

File: src/group/GroupCard.js [hunk 2]
```diff
@@ -103,7 +104,7 @@ export default class GroupCard extends PureComponent<Props, State> {
...
-          users={users}
+          users={users.filter(user => user.fullName.includes(filter))}
```
Comment (borisyankov) [inline src/group/GroupCard.js]: If you use 'filter' as @jainkuniya mentioned this will not be needed.

File: src/group/GroupScreen.js [hunk 3]
```diff
@@ -4,11 +4,24 @@ import React, { PureComponent } from 'react';
 import { Screen } from '../common';
 import GroupContainer from './GroupContainer';
 
-export default class GroupScreen extends PureComponent<{}> {
+type Props = {};
+type State = {
+  filter: string,
+};
+export default class GroupScreen extends PureComponent<Props, State> {
```
Comment (borisyankov) [inline src/group/GroupScreen.js]: Just include `{}` instead of `Props` here and remove `type Props = {}`.

File: src/nullObjects.js [hunk 4]
```diff
@@ -24,9 +24,13 @@ export const NULL_ACCOUNT: Account = {
 };
 
 export const NULL_USER: User = {
-  fullName: '',
   avatarUrl: '',
   email: '',
+  fullName: '',
```
Comment (borisyankov) [inline src/nullObjects.js]: This is a useful change. You can introduce it in a separate commit.

File: src/types.js [hunk 5]
```diff
@@ -72,7 +72,7 @@ export type User = any; /* {
   isActive: boolean,
   isAdmin: boolean,
   isBot: boolean,
-} */
+}; */
```
Comment (borisyankov) [inline src/types.js]: Careful not to change stuff you don't need to.

File: src/users/__tests__/userSelectors-test.js [hunk 6]
```diff
@@ -37,7 +37,15 @@ describe('getAccountDetailsUser', () => {
       },
       users: [],
     });
-    const expectedUser = { email: 'b@a.com', fullName: 'b@a.com', avatarUrl: '' };
+    const expectedUser = {
```
Comment (borisyankov) [inline src/users/__tests__/userSelectors-test.js]: Include this in a separate PR with the change to NULL_USER.

File: src/group/GroupCard.js [hunk 7]
```diff
@@ -49,9 +51,9 @@ export default class GroupCard extends PureComponent<Props, State> {
     const { selected } = this.state;
 
     const user = users.find(x => x.email === email);
-
+    const merge = [...selected, user];
     this.setState({
-      selected: [...selected, user],
+      selected: merge,
```
Comment (borisyankov) [inline src/group/GroupCard.js]: Why did you change that?

Comment (ebouJ) [reply src/group/GroupCard.js]: @borisyankov because I am trying to add a User type. Currently, there are 35 errors related to adding a user type, the change is honestly not relevant to this PR because I am trying to fix all the errors and this is one of them. "selected" is of type User, and "user" is not. For this reason, flow is throwing an error. This is an approach I tried t resolve this issue but it persisted so that is the same as what was there before. I will remove it from this PR because it's not relevant here and if I am done then I'll send a separate PR for that. I boiled down the errors to 20 so hopefully, soon I will send a separate PR for this. I am still learning how to use Github that's why because I didn't intend to include it in this PR.

File: src/users/__tests__/userSelectors-test.js (hunk 6 above)
Comment (ebouJ) [reply src/users/__tests__/userSelectors-test.js]: @borisyankov sure will do.

**Comment by ebouJ** [comment]
@borisyankov @jainkuniya  can you please guide me on this. I made all the requested change you have asked for on this Pull request. How do I update it with my recent changes?

**Comment by jainkuniya** [comment]
> How do I update it with my recent changes?

commit changes & push them to `Search` branch :)
`git push origin Search`

**Comment by ebouJ** [comment]
@borisyankov I updated this PR as requested.

File: src/group/GroupCard.js [hunk 8]
```diff
@@ -49,7 +51,6 @@ export default class GroupCard extends PureComponent<Props, State> {
     const { selected } = this.state;
 
     const user = users.find(x => x.email === email);
-
```
Comment (borisyankov) [inline src/group/GroupCard.js]: Make sure not to do unintended changes.

File: src/group/GroupScreen.js [hunk 9]
```diff
@@ -4,11 +4,23 @@ import React, { PureComponent } from 'react';
 import { Screen } from '../common';
 import GroupContainer from './GroupContainer';
 
-export default class GroupScreen extends PureComponent<{}> {
+type State = {
+  filter: string,
+};
+export default class GroupScreen extends PureComponent<{}, State> {
```
Comment (borisyankov) [inline src/group/GroupScreen.js]: Add an empty line before this one.

File: src/nullObjects.js (hunk 4 above)
Comment (borisyankov) [reply src/nullObjects.js]: Considering you did the change in the other PR, do remove this change even if it is not functionally different.

**Comment by borisyankov** [comment]
The PR looks OK from coding perspective, but be more careful with styles and unintended changes. Also fix conflicts with `master` branch.

File: src/group/GroupCard.js (hunk 8 above)
Comment (ebouJ) [reply src/group/GroupCard.js]: Thanks for the feedback. I will make that this will not repeat again. I am new to sending PR that's why but now I am okay with it, so I will avoid such from happening again.

**Comment by ebouJ** [comment]
@borisyankov thanks conflict resolved and I will be more careful about unintended changes.

**Comment by ebouJ** [comment]
@gnprice thanks. I rebase this branch with current master. This is one of my initial PR's so I wasn't familiar with git and the commit message guidelines which I am pretty comfortable with right now.

File: src/group/GroupScreen.js [hunk 10]
```diff
@@ -1,14 +1,25 @@
...
+  };
+
+  handleFilterChange = (filter: string) => this.setState({ filter });
+
   render() {
+    const { filter } = this.state;
     return (
-      <Screen title="New group chat">
-        <GroupContainer />
+      <Screen search searchBarOnChange={this.handleFilterChange} scrollView={false}>
```
Comment (borisyankov) [inline src/group/GroupScreen.js]: We no longer have the `scrollView` property.

**Comment by borisyankov** [comment]
@ebouJ from code perspective the PR looks ready to merge.
Remove the `scrollView` prop that we no longer use and go over the empty lines that you did add or remove and undo these changes.
Then I can merge this!

**Comment by ebouJ** [comment]
@borisyankov thanks. I fix it.

**Comment by ebouJ** [comment]
@borisyankov this PR is also ready.

**Comment by borisyankov** [comment]
Looks good!
